hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataResolvers.java

```diff
@@ -116,38 +115,45 @@ public List<MappingField> resolveAndValidateFields(
             }
         }
 
-        Stream<MappingField> keyFields = findMetadataResolver(options, true)
-                .resolveAndValidateFields(true, userFields, options, ss)
-                .filter(field -> !field.name().equals(KEY) || field.externalName().equals(KEY));
-        Stream<MappingField> valueFields = findMetadataResolver(options, false)
-                .resolveAndValidateFields(false, userFields, options, ss)
-                .filter(field -> !field.name().equals(VALUE) || field.externalName().equals(VALUE));
-
-        final TypeKind keyKind = TypesUtils.formatToTypeKind(getFormat(options, true));
-        if (NESTED_FIELDS_SUPPORTED_FORMATS.contains(keyKind)) {
-            keyFields = keyFields
-                    .peek(mappingField -> TypesUtils.enrichMappingFieldType(keyKind, mappingField, relationsStorage));
-        }
-
-        final TypeKind valueKind = TypesUtils.formatToTypeKind(getFormat(options, false));
-        if (NESTED_FIELDS_SUPPORTED_FORMATS.contains(valueKind)) {
-            valueFields = valueFields
-                    .peek(mappingField -> TypesUtils.enrichMappingFieldType(valueKind, mappingField, relationsStorage));
-        }
+        Stream<MappingField> keyFields = resolveAndValidateFields(true, userFields, options,
+                serializationService, relationsStorage);
+        Stream<MappingField> valueFields = resolveAndValidateFields(false, userFields, options,
+                serializationService, relationsStorage);
 
-        Map<String, MappingField> fields = concat(keyFields, valueFields)
+        Map<String, MappingField> fields = Stream.concat(keyFields, valueFields)
                 .collect(LinkedHashMap::new, (map, field) -> map.putIfAbsent(field.name(), field), Map::putAll);
 
         if (fields.isEmpty()) {
             throw QueryException.error("The resolved field list is empty");
         }
-
         return new ArrayList<>(fields.values());
     }
 
+    private Stream<MappingField> resolveAndValidateFields(
+            boolean isKey,
+            List<MappingField> userFields,
+            Map<String, String> options,
+            InternalSerializationService serializationService,
+            RelationsStorage relationsStorage
+    ) {
+        String format = getFormat(options, isKey);
+        if (format != null && NESTED_FIELDS_SUPPORTED_FORMATS.contains(format)) {
+            List<MappingField> fieldsWithCustomTypes = extractFields(userFields, isKey).values().stream()
+                    .filter(mappingField -> mappingField.type().isCustomType()).collect(toList());
+            if (!fieldsWithCustomTypes.isEmpty()) {
+                FieldEnricher<?, ?> enricher = getFieldEnricher(format, serializationService, relationsStorage);
+                fieldsWithCustomTypes.forEach(mappingField -> enricher.enrich(mappingField, options, isKey));
+            }
+        }
+
+        String name = isKey ? KEY : VALUE;
+        return findMetadataResolver(options, isKey)
```

Comment:
as per https://github.com/hazelcast/hazelcast/pull/25269#discussion_r1367732778 null is allowed, but still this will make the order of operations less surprising

Reply:
From https://github.com/hazelcast/hazelcast/pull/25269#discussion_r1367732778
> `KvMetadataResolver.resolveAndValidateFields()` needs to validate nested types recursively, and to be able to do this, it requires custom types to be resolved already.

Reply:
ok, we can keep it as is